File: AsmWrite.c
/* Mite assembly writer
 * Reuben Thomas    24/11/00-20/4/01 */


#include <stdint.h>
#include <string.h>
#include <ctype.h>
#include <math.h>
#include <limits.h>

#include <rrt/except.h>
#include <rrt/memory.h>

#include "Translate.h"


/* maximum number of octal digits in a uintptr_t */
#define SIZE_T_MAXLEN (sizeof(uintptr_t) * CHAR_BIT / 3)

/* Instructions indexed by opcode */
static const char *inst[] = {
  "", /* opcodes start at 0x01 */
  "lab",   "mov",   "movi",  "ldl",   "ld",    "st",
  "gets",  "sets",  "pop",   "push",  "add",   "sub",
  "mul",   "div",   "rem",   "and",   "or",    "xor",
  "sl",    "srl",   "sra",   "teq",   "tlt",   "tltu",
  "b",     "br",    "bf",    "bt",    "call",  "callr",
  "ret",   "calln", "lit",   "litl",  "space"
};

static void
putChar(Translator *t, char c)
{
  bufExt(t->wImg, t->wSize, t->wPtr - t->wImg + 1U);
  *t->wPtr++ = c;
}

static void
putStr(Translator *t, const char *s, uintptr_t len)
{
  bufExt(t->wImg, t->wSize, t->wPtr - t->wImg + len);
  memcpy(t->wPtr, s, len);
  t->wPtr += len;
}

static uintptr_t
writeNum(char *s, uintptr_t n)
{
  uintptr_t last = n ? log10(n) : 0;

  s += last;
  do {
    *s-- = '0' + n % 10;
    n /= 10;
  } while (n);
  return last + 1;
}

static void
putNum(Translator *t, uintptr_t n)
{
  bufExt(t->wImg, t->wSize, t->wPtr - t->wImg + SIZE_T_MAXLEN);
  t->wPtr += writeNum(t->wPtr, n);
}

static void
putLabTy(Translator *t, unsigned int ty)
{
  switch (ty) {
    case LABEL_B: putChar(t, 'b');  return;
    case LABEL_S: putChar(t, 's');  return;
    case LABEL_D: putChar(t, 'd');  return;
  }
}

static void
putLab(Translator *t, Label *l)
{
  putChar(t, ' ');
  putLabTy(t, l->ty);
  putNum(t, t->reader->labelMap(t, l).n);
}

static void
putLabDangle(Translator *t, Label *l)
{
  putLabTy(t, l->ty);
  addDangle(t, l);
}

static void
putImm(Translator *t, Immediate *i)
{
  if (i->e) putChar(t, 'e');
  if (i->s) putChar(t, 's');
  if (i->w) putChar(t, 'w');
  if (i->n) putChar(t, '-');
  putNum(t, i->v);
  if (i->r) {
    putStr(t, ">>", 2);
    putNum(t, i->r);
  }
}

static void
putOp(Translator *t, MiteValue op, unsigned int ty)
{
  switch (ty) {
    case op_r: putNum(t, op.r); break;
    case op_l: putLabTy(t, op.l->ty); putLabDangle(t, op.l); break;
    case op_L: putLabTy(t, op.l->ty); putLab(t, op.l); break;
    case op_b: putLabDangle(t, op.l); break;
    case op_s: putLabDangle(t, op.l); break;
    case op_d: putLabDangle(t, op.l); break;
    case op_i: putImm(t, op.i); break;
  }
}

static void
putInst(Translator *t, unsigned int i, MiteValue op1, MiteValue op2,
	MiteValue op3)
{
  unsigned int ops;
  
  putStr(t, inst[i], strlen(inst[i]));
  putChar(t, ' ');
  ops = opType[i];
  if (OP1(ops))
    putOp(t, op1, OP1(ops));
  if (OP2(ops)) {
    putChar(t, ' ');
    putOp(t, op2, OP2(ops));
  }
  if (OP3(ops)) {
    putChar(t, ' ');
    putOp(t, op3, OP3(ops));
  }
  putChar(t, '\n');
}

static uintptr_t
insertNum(Byte **s, uintptr_t n)
{
  *s += writeNum((char *)*s, n);
  return 0;
}

static void
resolve(Translator *t)
{
  Byte *fImg;
  Dangle *d;
  uintptr_t n;
 
  for (d = t->dangles->next, n = 0; d; d = d->next, n++);
  fImg = excMalloc(t->wPtr - t->wImg + n * SIZE_T_MAXLEN);
  insertDangles(t, fImg, fImg, insertNum);
}

File: src/asmWrite.lua
-- Mite assembly writer
-- (c) Reuben Thomas 2000

return Writer(
  "Mite object",
  [[
#include <stdint.h>
#include <string.h>
#include <ctype.h>
#include <math.h>
#include <limits.h>

#include "except.h"
#include "buffer.h"
#include "translate.h"


/* maximum number of octal digits in a Word
   (upper bound on max. no. of decimal digits) */
#define WORD_MAXLEN (sizeof(Word) * CHAR_BIT / 3)

static void
putChar(TState *t, char c)
{
  *t->wPtr++ = c;
}

static void
putStr(TState *t, const char *s, uintptr_t len)
{
  ensure(len);
  memcpy(t->wPtr, s, len);
  t->wPtr += len;
}

static uintptr_t
writeNum(Byte *s, uintptr_t n)
{
  uintptr_t last = n ? log10(n) : 0;
  s += last;
  do {
    *s-- = '0' + n % 10;
    n /= 10;
  } while (n);
  return last + 1;
}

static void
putNum(TState *t, uintptr_t n)
{
  t->wPtr += writeNum(t->wPtr, n);
}

static void
putLabTy(TState *t, LabelType ty)
{
  switch (ty) {
  case LABEL_B:
    putChar(t, 'b');
    return;
  case LABEL_S:
    putChar(t, 's');
    return;
  case LABEL_D:
    putChar(t, 'd');
    return;
  }
}

#define S(s) putStr(t, s, sizeof(s) / sizeof(char) - 1)

static void
putImm(TState *t, int f, int n, uintptr_t v, uintptr_t r)
{
  if (f & FLAG_E)
    putChar(t, 'e');
  if (f & FLAG_S)
    putChar(t, 's');
  if (f & FLAG_W)
    putChar(t, 'w');
  if (n)
    putChar(t, '-');
  putNum(t, v);
  if (r) {
    S(">>");
    putNum(t, r);
  }
}

#define NL \
  *(t->wPtr)++ = '\n'

#define SP \
  *(t->wPtr)++ = ' '

#define R(r) \
  SP; \
  putNum(t, r)

#define LabTy(L) \
  SP; \
  putLabTy(t, L)

#define Lab(ty, l) \
  SP; \
  putLabTy(t, ty); \
  addDangle(t, ty, l)

#define Imm(f, n, v, r) \
  SP; \
  putImm(t, f, n, v, r)


#define putInt(t, sgn, n) \
  t->wPtr += writeInt(t->wPtr, sgn, n)
#define putUInt(t, n) \
  putInt(t, 0, n)

static uintptr_t
writeUInt(Byte **s, uintptr_t n)
{
  *s += writeNum(*s, n);
  return 0;
}

#define DANGLE_MAXLEN WORD_MAXLEN
#define RESOLVE_IMG NULL
#define RESOLVE_PTR NULL
  ]],
  {
    Inst("lab",    "S(\"lab\"); LabTy(t1); Lab(t1, l2); NL;"),
    Inst("mov",    "S(\"mov\"); R(r1); R(r2); NL;"),
    Inst("movi",   "S(\"movi\"); R(r1); " ..
                   "Imm(i2_f, i2_sgn, i2_v, i2_r); NL;"),
    Inst("ldl",    "S(\"ldl\"); R(r1); Lab(LABEL_D, l2); NL;"),
    Inst("ld",     "S(\"ld\"); R(r1); R(r2); NL;"),
    Inst("st",     "S(\"st\"); R(r1); R(r2); NL;"),
    Inst("gets",   "S(\"gets\"); R(r1); NL;"),
    Inst("sets",   "S(\"sets\"); R(r1); NL;"),
    Inst("pop",    "S(\"pop\"); R(r1); NL;"),
    Inst("push",   "S(\"push\"); R(r1); NL;"),
    Inst("add",    "S(\"add\"); R(r1); R(r2); R(r3); NL;"),
    Inst("sub",    "S(\"sub\"); R(r1); R(r2); R(r3); NL;"),
    Inst("mul",    "S(\"mul\"); R(r1); R(r2); R(r3); NL;"),
    Inst("div",    "S(\"div\"); R(r1); R(r2); R(r3); NL;"),
    Inst("rem",    "S(\"rem\"); R(r1); R(r2); R(r3); NL;"),
    Inst("and",    "S(\"and\"); R(r1); R(r2); R(r3); NL;"),
    Inst("or",     "S(\"or\"); R(r1); R(r2); R(r3); NL;"),
    Inst("xor",    "S(\"xor\"); R(r1); R(r2); R(r3); NL;"),
    Inst("sl",     "S(\"sl\"); R(r1); R(r2); R(r3); NL;"),
    Inst("srl",    "S(\"srl\"); R(r1); R(r2); R(r3); NL;"),
    Inst("sra",    "S(\"sra\"); R(r1); R(r2); R(r3); NL;"),
    Inst("teq",    "S(\"teq\"); R(r1); R(r2); R(r3); NL;"),
    Inst("tlt",    "S(\"tlt\"); R(r1); R(r2); R(r3); NL;"),
    Inst("tltu",   "S(\"tltu\"); R(r1); R(r2); R(r3); NL;"),
    Inst("b",      "S(\"b\"); Lab(LABEL_B, l1); NL;"),
    Inst("br",     "S(\"br\"); R(r1); NL;"),
    Inst("bf",     "S(\"bf\"); R(r1); Lab(LABEL_B, l2); NL;"),
    Inst("bt",     "S(\"bt\"); R(r1); Lab(LABEL_B, l2); NL;"),
    Inst("call",   "S(\"call\"); Lab(LABEL_S, l1); NL;"),
    Inst("callr",  "S(\"callr\"); R(r1); NL;"),
    Inst("ret",    "S(\"ret\"); NL;"),
    Inst("calln",  "S(\"calln\"); R(r1); NL;"),
    Inst("lit",    "S(\"lit\"); " ..
                   "Imm(i1_f, i1_sgn, i1_v, i1_r); NL;"),
    Inst("litl",   "S(\"litl\"); Lab(t1, l2); NL;"),
    Inst("space",  "S(\"space\"); " ..
                   "Imm(i1_f, i1_sgn, i1_v, i1_r); NL;"), 
  }
)
